pinot-minion/src/main/java/org/apache/pinot/minion/event/MinionEventObservers.java

```diff
@@ -0,0 +1,138 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.minion.event;
+
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+import org.apache.pinot.minion.MinionConf;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class MinionEventObservers {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MinionEventObservers.class);
+
+  private static final MinionEventObservers DEFAULT_INSTANCE = new MinionEventObservers();
+  private static volatile MinionEventObservers _customInstance = null;
+
+  private final Map<String, MinionEventObserver> _taskEventObservers = new ConcurrentHashMap<>();
+  // Tasks are added roughly in time order, so use LinkedList instead of PriorityQueue for simplicity.
+  private final Queue<EndedTask> _endedTaskQueue = new LinkedList<>();
+  private final ReentrantLock _queueLock = new ReentrantLock();
+  private final Condition _availableToClean = _queueLock.newCondition();
```

Comment:
I feel that we can remove the synchronization logic by using (1) use `ConcurrentLinkedList`, and (2) run clean up periodically (since the clean up time does not need to be accurate).

Reply:
was mainly looking to use condition to avoid unnecessary runs; and wanted to reuse the CachedThreadPool.